### PR TITLE
Type operator spacing rule

### PR DESCRIPTION
**Pull request by RyanCavanaugh**
Enforces spaces around `|` and `&`, e.g. `string|number` is an error but `string | number` is not

### CODE REVIEW

**Comment by sandersn** [comment]
Does this already exist for binary operators, eg with `symbol.flags & SymbolFlags.Class`? It seems like the rule should be the same there.

**Comment by RyanCavanaugh** [comment]
For expression-level operators the rules are different - for example, `||` will often have its right operand on the next line, and `,` shouldn't have a space after its left operand. Union or intersection types should be more standardized.

**Comment by weswigham** [comment]
:+1:

**Comment by sandersn** [comment]
OK, makes sense. :+1:

File: scripts/tslint/typeOperatorSpacingRule.ts [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    public static FAILURE_STRING = "Place spaces around the '|' and '&' type operators";
+
+    public apply(sourceFile: ts.SourceFile): Lint.RuleFailure[] {
+        return this.applyWithWalker(new TypeOperatorSpacingWalker(sourceFile, this.getOptions()));
+    }
+}
+
+class TypeOperatorSpacingWalker extends Lint.RuleWalker {
+    public visitNode(node: ts.Node) {
+        if(node.kind === ts.SyntaxKind.UnionType || node.kind === ts.SyntaxKind.IntersectionType) {
```
Comment (DanielRosenwasser) [inline scripts/tslint/typeOperatorSpacingRule.ts]: Space after `if`

File: scripts/tslint/typeOperatorSpacingRule.ts [hunk 2]
```diff
@@ -0,0 +1,28 @@
+/// <reference path="../../node_modules/tslint/typings/typescriptServices.d.ts" />
+/// <reference path="../../node_modules/tslint/lib/tslint.d.ts" />
+
+
+export class Rule extends Lint.Rules.AbstractRule {
+    public static FAILURE_STRING = "Place spaces around the '|' and '&' type operators";
```
Comment (DanielRosenwasser) [inline scripts/tslint/typeOperatorSpacingRule.ts]: a single space

File: scripts/tslint/typeOperatorSpacingRule.ts [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    }
+}
+
+class TypeOperatorSpacingWalker extends Lint.RuleWalker {
+    public visitNode(node: ts.Node) {
+        if(node.kind === ts.SyntaxKind.UnionType || node.kind === ts.SyntaxKind.IntersectionType) {
+            let typeNode = <ts.UnionOrIntersectionTypeNode>node;
+            let expectedStart = typeNode.types[0].end + 2; // space, | or &
+            for (let i = 1; i < typeNode.types.length; i++) {
+                if(expectedStart !== typeNode.types[i].pos || typeNode.types[i].getLeadingTriviaWidth() !== 1) {
```
Comment (DanielRosenwasser) [inline scripts/tslint/typeOperatorSpacingRule.ts]: Space after `if`

File: scripts/tslint/typeOperatorSpacingRule.ts [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+    public apply(sourceFile: ts.SourceFile): Lint.RuleFailure[] {
+        return this.applyWithWalker(new TypeOperatorSpacingWalker(sourceFile, this.getOptions()));
+    }
+}
+
+class TypeOperatorSpacingWalker extends Lint.RuleWalker {
+    public visitNode(node: ts.Node) {
+        if(node.kind === ts.SyntaxKind.UnionType || node.kind === ts.SyntaxKind.IntersectionType) {
+            let typeNode = <ts.UnionOrIntersectionTypeNode>node;
```
Comment (DanielRosenwasser) [inline scripts/tslint/typeOperatorSpacingRule.ts]: You only refer to `typeNode.types`, so just create `constituentTypes`

File: scripts/tslint/typeOperatorSpacingRule.ts (hunk 3 above)
Comment (DanielRosenwasser) [inline scripts/tslint/typeOperatorSpacingRule.ts]: Add a `currentType`

File: scripts/tslint/typeOperatorSpacingRule.ts (hunk 2 above)
Comment (RyanCavanaugh) [inline scripts/tslint/typeOperatorSpacingRule.ts]: :sparkles: :bike: :house: :sparkles:

**Comment by DanielRosenwasser** [comment]
For the record, we might have to relax this rule down the road, but until then :+1:

**Pull request merged**